http-server-netty/src/main/java/io/micronaut/http/server/netty/ServerNettyCustomizer.java

```diff
@@ -0,0 +1,117 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.server.netty;
+
+import io.micronaut.core.annotation.NonNull;
+import io.netty.channel.Channel;
+
+/**
+ * Interface implemented by users to hook into the pipeline setup of the netty HTTP server.
+ * <br>
+ * The internal pipeline of our HTTP server is by necessity complex and unstable across versions.
+ * While we strive to retain compatibility, please do not make too many assumptions about the
+ * Micronaut HTTP server pipeline structure when implementing this interface.
+ * <br>
+ * Implementations of this interface are scoped to different lifetimes. The root customizers can be
+ * added through the {@link Registry}, and live while the server is running. Customizers with
+ * narrower scope are created from the root customizers through
+ * {@link #specializeForChannel specialization} when new netty channels are created by the
+ * server, e.g. when a client connects. These specialized customizers will then receive
+ * notifications when a certain step in the pipeline setup is reached, so they can do their own
+ * customizations on the channel and the channel pipeline they received in the specialization step.
+ *
+ * @since 3.6.0
+ * @author yawkat
+ */
+public interface ServerNettyCustomizer {
+    /**
+     * Specialize this customizer for the given channel. In the "boring" case, the customizer will
+     * first be specialized for a {@link io.netty.channel.socket.ServerSocketChannel} when the
+     * server binds to a configured TCP port, and then to a
+     * {@link io.netty.channel.socket.SocketChannel} when a client connects to that port.
+     * <br>
+     * However, there are a lot of weird cases where the actual behavior might diverge from this.
+     * For example, the channel types will different if the server is bound to a unix domain socket
+     * ({@link io.netty.channel.unix.ServerDomainSocketChannel} and
+     * {@link io.netty.channel.unix.DomainSocketChannel} respectively). In the case of an
+     * {@link io.netty.channel.embedded.EmbeddedChannel} used for testing, there might not be a
+     * listener channel at all. For HTTP/2, each HTTP stream may get its own channel that is
+     * specialized from the overall connection channel. And finally, HTTP/3 support has to use
+     * datagram channels instead of the socket-based ones.
+     *
+     * @param channel The new channel to specialize for.
+     * @param role The role (or scope) of the channel.
+     * @return The new customizer, or {@code this} if no specialization needs to take place.
+     */
+    @NonNull
+    default ServerNettyCustomizer specializeForChannel(@NonNull Channel channel, @NonNull ChannelRole role) {
```

Comment:
does it make sense to future this API by adding a context parameter that allows looking up the channel and role? This allows adding other lookup methods to the context object without introducing a new method signature

Reply:
i think in that case i would just add a new specialize method, such as `default ServerNettyCustomizer specializeForX() { return this; }`